### PR TITLE
#8492 #8513 - Add a FuelCellElectricGenerator Unit Test and do not zero out Constant Skin Loss Rate in Generator:FuelCell:PowerModule

### LINKED ISSUE #8492: Unfinished FuelCellElectricGenerator Unit Test
Issue overview
--------------
The FuelCellElectricGenerator unit test is unfinished. 

https://github.com/NREL/EnergyPlus/blob/18f1ce3d49ae10de396d35cb18ceb0735c93e9c8/tst/EnergyPlus/unit/FuelCellElectricGenerator.unit.cc

Someone should flesh this out so that it at least processes the input.

### Details

Some additional details for this issue (if relevant):
 - Platform (Operating system, version): all
 - Version of EnergyPlus (if using an intermediate build, include SHA): 9.4+, 18f1ce3d49ae10de396d35cb18ceb0735c93e9c8

### Checklist

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### LINKED ISSUE #8513: Generator:FuelCell:PowerModule: Constant Skin Loss Rate overriden to zero
Issue overview
--------------

Found while adding a test in https://github.com/NREL/EnergyPlus/pull/8511

When `Skin Loss Calculation Mode == ConstantRate`, I expect the value of `Constant Skin Loss Rate {W}` to be used, but it's overriden to zero.

```
Generator:FuelCell:PowerModule,
  Generator Fuel Cell Power Module 1,     !- Name
  Annex42,                                !- Efficiency Curve Mode
  Power Module Efficiency Curve,          !- Efficiency Curve Name
  1,                                      !- Nominal Efficiency
  3400,                                   !- Nominal Electrical Power {W}
  10,                                     !- Number of Stops at Start of Simulation
  0.1,                                    !- Cycling Performance Degradation Coefficient
  0,                                      !- Number of Run Hours at Beginning of Simulation {hr}
  0,                                      !- Accumulated Run Time Degradation Coefficient
  10000,                                  !- Run Time Degradation Initiation Time Threshold {hr}
  1.4,                                    !- Power Up Transient Limit {W/s}
  0.2,                                    !- Power Down Transient Limit {W/s}
  0,                                      !- Start Up Time {s}
  0.2,                                    !- Start Up Fuel {kmol}
  0,                                      !- Start Up Electricity Consumption {J}
  0,                                      !- Start Up Electricity Produced {J}
  0,                                      !- Shut Down Time {s}
  0.2,                                    !- Shut Down Fuel {kmol}
  0,                                      !- Shut Down Electricity Consumption {J}
  0,                                      !- Ancillary Electricity Constant Term
  0,                                      !- Ancillary Electricity Linear Term
  ConstantRate,                           !- Skin Loss Calculation Mode
  Thermal Zone 1,                         !- Zone Name
  0.6392,                                 !- Skin Loss Radiative Fraction
  729,                                    !- Constant Skin Loss Rate {W}
  0,                                      !- Skin Loss U-Factor Times Area Term {W/K}
  Skin Loss Curve,                        !- Skin Loss Quadratic Curve Name
  0.006156,                               !- Dilution Air Flow Rate {kmol/s}
  2307,                                   !- Stack Heat loss to Dilution Air {W}
  Generator Fuel Cell Power Module 1 OA Node, !- Dilution Inlet Air Node Name
  Generator Fuel Cell Power Module 1 Dilution Outlet Air Node, !- Dilution Outlet Air Node Name
  3010,                                   !- Minimum Operating Point {W}
  3728;                                   !- Maximum Operating Point {W}
```

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version): all
 - Version of EnergyPlus (if using an intermediate build, include SHA): 9.4.0+, develop at  f754bc5c2c5d415789c9b01297c575772833b862

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### LINKED ISSUE #8514: Generator:FuelCell:PowerModule: Number of Stops at Start of Simulation is ignored
Issue overview
--------------

Found while adding a test in #8511. NumCycles is zero, even though I entered "10".

```
Generator:FuelCell:PowerModule,
  Generator Fuel Cell Power Module 1,     !- Name
  Annex42,                                !- Efficiency Curve Mode
  Power Module Efficiency Curve,          !- Efficiency Curve Name
  1,                                      !- Nominal Efficiency
  3400,                                   !- Nominal Electrical Power {W}
  10,                                     !- Number of Stops at Start of Simulation
  0.1,                                    !- Cycling Performance Degradation Coefficient
  0,                                      !- Number of Run Hours at Beginning of Simulation {hr}
  0,                                      !- Accumulated Run Time Degradation Coefficient
  10000,                                  !- Run Time Degradation Initiation Time Threshold {hr}
  1.4,                                    !- Power Up Transient Limit {W/s}
  0.2,                                    !- Power Down Transient Limit {W/s}
  0,                                      !- Start Up Time {s}
  0.2,                                    !- Start Up Fuel {kmol}
  0,                                      !- Start Up Electricity Consumption {J}
  0,                                      !- Start Up Electricity Produced {J}
  0,                                      !- Shut Down Time {s}
  0.2,                                    !- Shut Down Fuel {kmol}
  0,                                      !- Shut Down Electricity Consumption {J}
  0,                                      !- Ancillary Electricity Constant Term
  0,                                      !- Ancillary Electricity Linear Term
  ConstantRate,                           !- Skin Loss Calculation Mode
  Thermal Zone 1,                         !- Zone Name
  0.6392,                                 !- Skin Loss Radiative Fraction
  729,                                    !- Constant Skin Loss Rate {W}
  0,                                      !- Skin Loss U-Factor Times Area Term {W/K}
  Skin Loss Curve,                        !- Skin Loss Quadratic Curve Name
  0.006156,                               !- Dilution Air Flow Rate {kmol/s}
  2307,                                   !- Stack Heat loss to Dilution Air {W}
  Generator Fuel Cell Power Module 1 OA Node, !- Dilution Inlet Air Node Name
  Generator Fuel Cell Power Module 1 Dilution Outlet Air Node, !- Dilution Outlet Air Node Name
  3010,                                   !- Minimum Operating Point {W}
  3728;                                   !- Maximum Operating Point {W}
```

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version): all
 - Version of EnergyPlus (if using an intermediate build, include SHA): 9.4.0+, develop at f754bc5

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------
 - Fixes #8492 - Add a FuelCellElectricGenerator Unit Test
 - Fixes #8513 - Generator:FuelCell:PowerModule: Constant Skin Loss Rate overriden to zero
 - Fixes #8514 - Generator:FuelCell:PowerModule: Number of Stops at Start of Simulation is ignored

**Diffs are explained below**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [x] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [x] If structural output changes, add to output rules file and add OutputChange label
 - [x] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: tst/EnergyPlus/unit/FuelCellElectricGenerator.unit.cc [hunk 1]
```diff
@@ -55,132 +55,1027 @@
...
+    EXPECT_EQ(DataGenerators::SkinLoss::ConstantRate, fCPM.SkinLossMode);
+
+    EXPECT_EQ("THERMAL ZONE 1", fCPM.ZoneName);
+    EXPECT_EQ(1, fCPM.ZoneID);
+
+    EXPECT_EQ(0.6392, fCPM.RadiativeFract);
+
+    // TODO: this is an actual bug. It is being overriden in FigureFuelCellZoneGains, while here the SkinLossMode is ConstantRate, so it should be
+    // unaffected throughout
+    EXPECT_EQ(729, fCPM.QdotSkin);
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/FuelCellElectricGenerator.unit.cc]: TODO: this seems like a bug.

It's being overriden to 0.0 here: https://github.com/NREL/EnergyPlus/blob/f754bc5c2c5d415789c9b01297c575772833b862/src/EnergyPlus/FuelCellElectricGenerator.cc#L3343

File: tst/EnergyPlus/unit/FuelCellElectricGenerator.unit.cc [hunk 2]
```diff
@@ -55,132 +55,1027 @@
...
+    auto airSup = thisFC->AirSup;
+    EXPECT_EQ(airSup.Name, thisFC->NameFCAirSup);
+    EXPECT_EQ("GENERATOR FUEL CELL AIR SUPPLY 1 OA NODE", airSup.NodeName);
+    ASSERT_GT(airSup.BlowerPowerCurveID, 0);
+    EXPECT_EQ("BLOWER POWER CURVE", state->dataCurveManager->PerfCurve(airSup.BlowerPowerCurveID).Name);
+    EXPECT_EQ(1.0, airSup.BlowerHeatLossFactor);
+    EXPECT_EQ(DataGenerators::AirSupRateMode::ConstantStoicsAirRat, airSup.AirSupRateMode);
+
+    // TODO: Interestingly, Stoics ratio is the input + 1.0
+    EXPECT_EQ(2.0, airSup.Stoics);
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/FuelCellElectricGenerator.unit.cc]: haven't really had time to track how that makes sense, but will do.

Comment (jmarrec) [reply tst/EnergyPlus/unit/FuelCellElectricGenerator.unit.cc]: Well, at least the IO ref mentions it, so no problem: https://bigladdersoftware.com/epx/docs/9-4/input-output-reference/group-electric-load-center-generator.html#field-stoichiometric-ratio

File: tst/EnergyPlus/unit/FuelCellElectricGenerator.unit.cc [hunk 3]
```diff
@@ -55,132 +55,1027 @@
...
+        "  0,                                      !- Start Up Electricity Produced {J}",
+        "  0,                                      !- Shut Down Time {s}",
+        "  0.2,                                    !- Shut Down Fuel {kmol}",
+        "  0,                                      !- Shut Down Electricity Consumption {J}",
+        "  0,                                      !- Ancillary Electricity Constant Term",
+        "  0,                                      !- Ancillary Electricity Linear Term",
+        "  ConstantRate,                           !- Skin Loss Calculation Mode",
+        "  Thermal Zone 1,                         !- Zone Name",
+        "  0.6392,                                 !- Skin Loss Radiative Fraction",
+        "  729,                                    !- Constant Skin Loss Rate {W}",
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/FuelCellElectricGenerator.unit.cc:407]: FCMP.QdotSkin = 729

File: tst/EnergyPlus/unit/FuelCellElectricGenerator.unit.cc (hunk 1 above)
Comment (jmarrec) [reply tst/EnergyPlus/unit/FuelCellElectricGenerator.unit.cc]: Addressed it in https://github.com/NREL/EnergyPlus/pull/8511/commits/4f837e59b1648cb73aeeb7af39dfb81ea96affcd

File: src/EnergyPlus/FuelCellElectricGenerator.cc [hunk 4]
```diff
@@ -266,7 +266,8 @@ namespace FuelCellElectricGenerator {
 
                 state.dataFuelCellElectGen->FuelCell(thisFuelCell).FCPM.NomEff = NumArray(1);
                 state.dataFuelCellElectGen->FuelCell(thisFuelCell).FCPM.NomPel = NumArray(2);
-                state.dataFuelCellElectGen->FuelCell(thisFuelCell).FCPM.NumCycles = NumArray(3);
+                state.dataFuelCellElectGen->FuelCell(thisFuelCell).FCPM.NumCyclesAtStart = NumArray(3);
+                state.dataFuelCellElectGen->FuelCell(thisFuelCell).FCPM.NumCycles = state.dataFuelCellElectGen->FuelCell(thisFuelCell).FCPM.NumCyclesAtStart;
```
Comment (jmarrec) [inline src/EnergyPlus/FuelCellElectricGenerator.cc:270]: Fix #8514 - Generator:FuelCell:PowerModule: Number of Stops at Start of Simulation is ignored

File: src/EnergyPlus/FuelCellElectricGenerator.cc [hunk 5]
```diff
@@ -3194,7 +3195,7 @@ namespace FuelCellElectricGenerator {
             this->AirSup.TairIntoBlower = 0.0;
             this->AirSup.QskinLoss = 0.0;
             this->AirSup.QintakeRecovery = 0.0;
-            this->FCPM.NumCycles = 0;
+            this->FCPM.NumCycles = this->FCPM.NumCyclesAtStart;
```
Comment (jmarrec) [inline src/EnergyPlus/FuelCellElectricGenerator.cc:3212]: Fix #8514 - Generator:FuelCell:PowerModule: Number of Stops at Start of Simulation is ignored

File: src/EnergyPlus/FuelCellElectricGenerator.cc [hunk 6]
```diff
@@ -3340,7 +3341,10 @@ namespace FuelCellElectricGenerator {
                 cell.AirSup.QskinLoss = 0.0;
                 cell.WaterSup.QskinLoss = 0.0;
                 cell.AuxilHeat.QskinLoss = 0.0;
-                cell.FCPM.QdotSkin = 0.0;
+                if (cell.FCPM.SkinLossMode != DataGenerators::SkinLoss::ConstantRate) {
+                    // If Constant Skin Loss Rate, then do not zero out
+                    cell.FCPM.QdotSkin = 0.0;
+                }
```
Comment (jmarrec) [inline src/EnergyPlus/FuelCellElectricGenerator.cc:3361]: Fix #8513

File: src/EnergyPlus/FuelCellElectricGenerator.cc [hunk 7]
```diff
@@ -1330,6 +1331,20 @@ namespace FuelCellElectricGenerator {
...
+                                "System",
+                                "Average",
+                                this->Name);
+
+            SetupOutputVariable(state, "Generator Power Module Skin Heat Loss Rate",
+                                OutputProcessor::Unit::W,
+                                this->Report.FCPMSkinLoss,
+                                "System",
+                                "Average",
+                                this->Name);
```
Comment (jmarrec) [inline src/EnergyPlus/FuelCellElectricGenerator.cc:1347]: Register two new **Advanced** report output variables to test changes for #8513 and #8514 

You need this to be able to see them:

```
  Output:Diagnostics,
    DisplayAdvancedReportVariables;
```

**Comment by jmarrec** [comment]
# Explaining CI diffs in MicroCogeneration

To explain the **expected differences** that CI reports for MicroCogeneration, I added a modified `MicroCogeneration.idf` on EnergyPlusDevSupport:

https://github.com/NREL/EnergyPlusDevSupport/blob/d9b935df34d66e70773ab003a2ddbddd09e002ab/DefectFiles/8000s/8492/MicroCogeneration.idf#L5170-L5182

I also cherry picked https://github.com/NREL/EnergyPlus/pull/8511/commits/ccaad9edc27e197758b974290562c16f5d6de124 onto develop, so I would have the new output variables I registered available.

I ran an annual simulation with both with output variables added, and here are the results.


## #8514 - FCPM Number of cycles

I set the FCPM `Number of Stops at Start of Simulation` to **10**.

https://github.com/NREL/EnergyPlusDevSupport/blob/d9b935df34d66e70773ab003a2ddbddd09e002ab/DefectFiles/8000s/8492/MicroCogeneration.idf#L4859

As you can see, before it was zero, now it's properly 10.

image

image


## #8513 FCPM Constant Skin Loss

The FCPM `Constant Skin Loss Rate {W}` is **729**: https://github.com/NREL/EnergyPlusDevSupport/blob/d9b935df34d66e70773ab003a2ddbddd09e002ab/DefectFiles/8000s/8492/MicroCogeneration.idf#L4878

Looking at the "Generator Zone Sensible Heat Transfer Rate", you can see that there is indeed 729 W accounted for now, that where missing before:

image


And using the specific advanced output variable I added, it's crystal clear:

image


image

**Comment by Myoldmopar** [comment]
This looks really good to me.  I'll build it locally with all the changes to develop since the last commit and we'll see if this can go right in.  Thanks @jmarrec

**Comment by dareumnam** [comment]
This looks really good to me too. I built it locally and ran the unit test successfully. I ran the example file and it runs cleanly, showed the result as @jmarrec mentioned in the comment above. I think the only thing that should be done before this gets merged in is adding the explanation about the two new output variables in I/O Reference Doc. Thank you @jmarrec

File: tst/EnergyPlus/unit/FuelCellElectricGenerator.unit.cc [hunk 8]
```diff
@@ -55,132 +55,970 @@
...
+
+    auto report = thisFC->Report;
+    EXPECT_EQ(exhaustHX.qHX, report.qHX);
+    EXPECT_EQ(report.ACPowerGen, generatorController->electProdRate);
+    EXPECT_EQ(report.ACEnergyGen, generatorController->electricityProd);
+
+    EXPECT_TRUE(generatorController->electProdRate > 1.15 * generatorController->thermProdRate) << "Power to Heat Ratio appears too low";
+    EXPECT_DOUBLE_EQ(exhaustHX.qHX, generatorController->thermProdRate);
+    EXPECT_DOUBLE_EQ(generatorController->thermProdRate * 15 * 60, generatorController->thermalProd);
 }
```
Comment (dareumnam) [inline tst/EnergyPlus/unit/FuelCellElectricGenerator.unit.cc:1024]: Thank you so much for finishing this huge unit test and catching two more defects about this while at it.  We appreciate it.

**Comment by jmarrec** [comment]
@dareumnam Thanks for the review and the catch about I/O ref. I added the variables in https://github.com/NREL/EnergyPlus/pull/8511/commits/2babab7296e646d8807ee440e245fce71424d8ba:

*  in the list of advanced report variables:

image

* And added a description with some links where appropriate and testing that the links do work correctly:

image

**Comment by dareumnam** [comment]
Doc looks great, CI is all happy here. I'll merge this now. Thanks so much @jmarrec for the fix.
